Task 90, updated code

dpisarenko opened the pull request:
Second pull request for fixing the issue https://github.com/yegor256/s3auth/issues/90 .

**Notes:**

1) Please not that in order to build the project, I need to use the 1.0-SNAPSHOT version of the jcabi-ssl-maven-plugin until the issue https://github.com/jcabi/jcabi-ssl-maven-plugin/issues/4 is fixed. I created a pull request with changes that will fix it.

2) I'm getting lots of code quality errors like shown below, which I'm fixing now.

```
[ERROR] \src\main\java\com\s3auth\hosts\DefaultDynamo.java[30]: Lines in file sh
ould end with Unix-like end of line (RegexpMultilineCheck)
[ERROR] \src\main\java\com\s3auth\hosts\DefaultDynamo.java[55]: Lines in file sh
ould end with Unix-like end of line (RegexpMultilineCheck)
[ERROR] \src\main\java\com\s3auth\hosts\DefaultDynamo.java[71]: Lines in file sh
ould end with Unix-like end of line (RegexpMultilineCheck)
```

3) The tests seem to run.

yegor256 commented:
@davvd please assign @carlosmiranda to this PR

davvd commented:
> @davvd please assign @carlosmiranda to this PR

@yegor256 sure! @carlosmiranda the task is yours, please go ahead

carlosmiranda commented:
@dpisarenko Have you tried doing a clean checkout with Unix-like line endings configured? See https://help.github.com/articles/dealing-with-line-endings/

carlosmiranda commented on pom.xml:
```diff
@@ -186,7 +186,13 @@
                 <artifactId>jcabi-ssl-maven-plugin</artifactId>
-                <version>0.7.22</version>
+                <!--
+                    I'm using 1.0-SNAPSHOT only in order to be able to build s3auth.
```
Let's format this as a puzzle, so that it can be picked up and made into a separate ticket to be fixed in the future. It's as simple as starting the first line of the comment with `@todo <issue number>:<estimate>`. For example:

```xml
<!--
    @todo #90:30min I'm using 1.0-SNAPSHOT only...
   ...
--->
```

See http://www.yegor256.com/2009/03/04/pdd.html for an explanation of how it works.

carlosmiranda commented on s3auth-hosts/pom.xml:
```diff
@@ -40,6 +40,7 @@
     <artifactId>s3auth-hosts</artifactId>
     <packaging>jar</packaging>
     <name>s3auth-hosts</name>
+
```
Let's remove this empty line, it's not necessary.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, URN of a user.
      */
-    public static final String USER = "user.urn";
+    public static final String USER = "user_urn";
```
We can't change the name of the Dynamo DB key, since that's how the current database exists. The tests should accommodate what's there, not the other way around.

dpisarenko replied:
@carlosmiranda jcabi-dynamo doesn't allow you to use periods in column names. I reported a bug - https://github.com/jcabi/jcabi-dynamo/issues/20 .

carlosmiranda replied:
@dpisarenko We're aware of the bug in jcabi/jcabi-dynamo#7. If you want we can wait for it, please ask @davvd and @yegor256 to put the task on hold. It's a mock framework problem, we still shouldn't change production to accommodate tests.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, name of domain.
      */
-    public static final String NAME = "domain.name";
+    public static final String NAME = "domain_name";
```
Same here.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, AWS key of bucket.
      */
-    public static final String KEY = "domain.key";
+    public static final String KEY = "domain_key";
```
Here also.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, Name of bucket.
      */
-    public static final String BUCKET = "domain.bucket";
+    public static final String BUCKET = "domain_bucket";
```
Same here.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, AWS S3 region of bucket.
      */
-    public static final String REGION = "domain.region";
+    public static final String REGION = "domain_region";
```
Same here.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -56,58 +58,58 @@
      * Dynamo DB key, Syslog host and port of domain.
      */
-    public static final String SYSLOG = "domain.syslog";
+    public static final String SYSLOG = "domain_syslog";
```
Same here, let's not change the key name.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -200,35 +205,74 @@ public void close() {
-        amazon.shutdown();
-        return true;
+        final Region region = this.regionfactory.createRegion();
+        final Table curTable = region.table(this.table);
```
Avoid composite variable names; something like `current` is sufficient. See http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/DefaultDynamo.java:
```diff
@@ -200,35 +205,74 @@ public void close() {
-        amazon.shutdown();
+        final Region region = this.regionfactory.createRegion();
+        final Table curTable = region.table(this.table);
+        final Iterator<Item> itemsToRemove = curTable.frame().where(
```
Same here, avoid composite variable names: http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/RegionCreator.java:
```diff
@@ -0,0 +1,54 @@
+ * @version $Id$
+ */
+@Immutable
+public interface RegionCreator {
```
Classes should be named as abstractions of real-life entities; i.e. by what they "are" not by what they "do". How about `Regions`? See http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/RegionCreator.java:
```diff
@@ -0,0 +1,54 @@
+     * Creates a region.
+     * @return Created region.
+     */
+    Region createRegion();
```
Let's call this method `create()`.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/ReRegionCreator.java:
```diff
@@ -0,0 +1,66 @@
+ * @author Dmitri Pisarenko (dp@altruix.co)
+ * @version $Id$
+ */
+public class ReRegionCreator implements RegionCreator {
```
Let's name this class `ReRegions` instead: http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#class-names

carlosmiranda replied:
No unit test for this class? Also, I'm not sure why we need this class and the `RegionCreator` interface, the whole thing seems like it's just passing data. Why not incorporate all this functionality in `DefaultDynamo`?

dpisarenko replied:
I added a test `com.s3auth.hosts.ReRegionsTest`. We need this class in order to be able to create mock regions (based on embedded H2 database instead of the real AWS DB) in `com.s3auth.hosts.DefaultDynamoTest#loadsDynamoConfiguration`. It's a way to make `DefaultDynamo` testable.

carlosmiranda commented:
@dpisarenko A few specific comments above. Furthermore, a couple of general comments:

1. Don't ignore the static analysis errors, our merging bot won't accept any changes if there are warnings or errors: http://www.yegor256.com/2014/08/13/strict-code-quality-control.html#first-reaction
2. The branch is too big. This may take some getting used to, but in PDD we don't necessarily finish everything in one go. You can choose to implement some stubs with corresponding ignored unit tests, and describe the remaining work in some puzzle comments. See http://www.yegor256.com/2009/03/04/pdd.html

dpisarenko commented:
@carlosmiranda Regarding line endings: `git config --list` says that `core.autocrlf=true` in my s3auth repository.

dpisarenko commented:
@carlosmiranda Now the buld fails with an error I don't know how to fix - `Found duplicate classes/resources`. If you know, how to diagnose/fix it, please tell me.

```
[ERROR] Failed to execute goal com.qulice:qulice-maven-plugin:0.12:check (jcabi-check) on project s3auth-hosts: Execution jcabi-check of goal com.qulice:qulice-maven-plugin:0.12:check failed: org.apache.maven.plugin.MojoExecutionException: Found duplicate classes/resources -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.qulice:qulice-maven-plugin:0.12:check (jcabi-check) on project s3auth-hosts: Execution jcabi-check of goal com.qulice:qulice-maven-plugin:0.12:check failed: org.apache.maven.plugin.MojoExecutionException: Found duplicate classes/resources
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:224)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:582)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:214)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:158)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution jcabi-check of goal com.qulice:qulice-maven-plugin:0.12:check failed: org.apache.maven.plugin.MojoExecutionException: Found duplicate classes/resources
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:143)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 19 more
Caused by: java.lang.IllegalArgumentException: org.apache.maven.plugin.MojoExecutionException: Found duplicate classes/resources
	at com.qulice.maven.MojoExecutor.execute(MojoExecutor.java:136)
	at com.qulice.maven.DuplicateFinderValidator.validate(DuplicateFinderValidator.java:91)
	at com.qulice.maven.CheckMojo.run(CheckMojo.java:86)
	at com.qulice.maven.CheckMojo.doExecute(CheckMojo.java:59)
	at com.qulice.maven.AbstractQuliceMojo.execute(AbstractQuliceMojo.java:175)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	... 20 more
Caused by: org.apache.maven.plugin.MojoExecutionException: Found duplicate classes/resources
	at com.ning.maven.plugins.duplicatefinder.DuplicateFinderMojo.checkClasspath(DuplicateFinderMojo.java:250)
	at com.ning.maven.plugins.duplicatefinder.DuplicateFinderMojo.checkCompileClasspath(DuplicateFinderMojo.java:201)
	at com.ning.maven.plugins.duplicatefinder.DuplicateFinderMojo.execute(DuplicateFinderMojo.java:178)
	at com.qulice.maven.MojoExecutor.execute(MojoExecutor.java:134)
	... 25 more
```

dpisarenko commented:
@carlosmiranda Note that the issue https://github.com/jcabi/jcabi-ssl-maven-plugin/issues/4 I mentioned above is now fixed and the PR was merged. But I don't know, how to make sure that the s3auth build uses the latest version. I added an issue for that ( https://github.com/jcabi/jcabi-ssl-maven-plugin/issues/12 ).

carlosmiranda commented:
@dpisarenko Regarding the line endings, `core.autocrlf` must be set to `false`, and you need to do a clean checkout.

carlosmiranda commented on s3auth-hosts/src/main/java/com/s3auth/hosts/ReRegions.java:
```diff
@@ -0,0 +1,66 @@
+ * @author Dmitri Pisarenko (dp@altruix.co)
+ * @version $Id$
+ */
+public class ReRegions implements Regions {
```
I still don't see the point of this class. Why not use `com.jcabi.dynamo.Region` instead of this class as a parameter `DefaultDynamo`'s constructor and pass `MkRegion` to it in unit tests? http://dynamo.jcabi.com/example-mock.html

carlosmiranda commented:
@dpisarenko Regarding your build failure, it would be better if you can raise a new ticket so that we can address it separately. Please see the following posts for guidance:

http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html
http://www.yegor256.com/2014/04/13/bugs-are-welcome.html

carlosmiranda commented:
@dpisarenko Do we have any updates here?

dpisarenko commented:
@carlosmiranda No. On January 5th I decided to stop working on the open tasks (including this one), asked Yegor whether I have to finalize them and he said no (that I don't have to finalize it). So I didn't.

I think someone else has been assigned to the issue in question since then.

carlosmiranda commented:
@dpisarenko No problem - can we close the pull request, then?

davvd commented:
@carlosmiranda I just added **27 mins** to your account, many thanks for your contribution..

+27 to your rating, your total score is [+968](http://www.netbout.com/b/35023?open=rating)
